plugins/experimental/carp/CarpConfigPool.cc

```diff
@@ -0,0 +1,218 @@
+  limitations under the License.
+ */
+#include <sstream>
+
+#include "Common.h"
+#include "CarpConfigPool.h"
+#include "UrlComponents.h"
+#include "CarpHost.h"
+
+#include <netdb.h>
+#include <ts/ts.h>
+ 
+using namespace std;
+
+/*******************************************************************/
+CarpConfigPool::CarpConfigPool()
+{
+  _globalHash = NULL;
+  _globalConfig = NULL;
+}
+
+/*******************************************************************/
+CarpConfigPool::~CarpConfigPool()
+{
+  for(CarpConfigListIter it=_configList.begin(); it != _configList.end(); it++) {
+    it->second->_config->stop();
+    delete it->second;
+  }
+}
+/*******************************************************************/
+static int initCarpConfigAndHash(CarpConfigAndHash * cch, string sFilename) {
+  cch->_configPath = sFilename;
+  cch->_config = new CarpConfig();
+  TSAssert(cch->_config);
+
+  if (!cch->_config->loadConfig(sFilename)) {
+    return -1;
+  }
+
+  cch->_hashAlgo = new CarpHashAlgorithm(cch->_config);
+
+  TSAssert(cch->_hashAlgo);
+
+  CarpHostList* hostList = cch->_config->getHostList();
+  // add hosts, etc to hash algo
+  char szServerName[256];
+  *szServerName = 0;
+
+  if (!gethostname(szServerName, 255)) { // success!
+    TSDebug(DEBUG_TAG_INIT, "using %s as server name to detect 'self'",
+        szServerName);
+  }
+  char buf[1024];
+  struct hostent self, *selfhe = getHostIp(string(szServerName), &self, buf,
+      sizeof(buf));
+
+  for (CarpHostListIter i = hostList->begin(); i != hostList->end(); i++) {
+    HashNode *hashNode;
+    bool bSelf = false;
+    if (NULL != selfhe) { // check for 'self'
+      // include hostname and port
+      bSelf = isSelf((*i)->getName(), (*i)->getPort(), selfhe);
+    }
+    if (cch->_config->getHealthCheckPort() == -1) {	// did they specify 'PORT'?
+      (*i)->setHealthCheckPort((*i)->getPort()); // set HC port from server spec'd port
+    } else {
+      (*i)->setHealthCheckPort(cch->_config->getHealthCheckPort()); // set HC port
+    }
+    string sHCUrl = cch->_config->getHealthCheckUrl();
+    size_t pos = sHCUrl.find("{port}");
+    if (pos != string::npos) { // need to replace '{port}' with servers port
+      stringstream ss;
+      ss << (*i)->getPort();
+      sHCUrl.replace(pos, 6, ss.str()); // 6 = strlen of '{port}'
+    }
+    pos = sHCUrl.find("{host}");
+    if (pos != string::npos) {
+      sHCUrl.replace(pos, 6, (*i)->getName());
+    }
+    (*i)->setHealthCheckUrl(sHCUrl); // set HC Url
+
+    // Look up host and create addr struct for healthchecks
+    char hBuf[1024];
+    struct hostent host, *hosthe = getHostIp((*i)->getName(), &host, hBuf,
+        sizeof(hBuf));
+    if (hosthe) {
+      // convert hostent to sockaddr_in structure
+      sockaddr_in hIn;
+      memcpy(&hIn.sin_addr, hosthe->h_addr_list[0], hosthe->h_length);
+      hIn.sin_port = htons((*i)->getHealthCheckPort()); // set port
+      if (hosthe->h_length == 4) { // size match IPv4? assume such
```

Comment:
If it's really an IP address the family member of the sockaddr should be checked, not the length.